### DISCUSSION TITLE
Progress bar between the routes

### DISCUSSION THREAD

**Discussion by eMeRiKa13** [nuxt/nuxt#17428] [Ideas]
I found in the Nuxt 2 documentation:

> "**The loading property**: Nuxt uses its own component to show a progress bar between the routes."

I don't think it exists yet with Nuxt3? Is-it something planned for the future? Or any recommandation of a good third party solution?

Thanks

**Comment by lacorde** [comment]
Would also like to know !

**Comment by aligzl** [comment]
I figure out it by fallowing code and nuxt-loading component ( refferanced from nuxt-2 ) in app.vue

**app.vue**
```

<template>
  <div class="scroll-smooth">
    <NuxtPage/>
    <nuxt-loading ref="loading" />
  </div>
</template>
<script setup>

  const loading = ref(null)

  const nuxtApp = useNuxtApp()

  nuxtApp.hook("page:start", () => {
    loading.value.start()
  })

  nuxtApp.hook("page:finish", () => {
    loading.value.finish()
     window.scrollTo(0, 0)
  })
</script>
```
**nuxt-loading.vue**

```

<script>
export default {
  name: 'NuxtLoading',
  data () {
    return {
      percent: 0,
      show: false,
      canSucceed: true,
      reversed: false,
      skipTimerCount: 0,
     
      throttle: 200,
      duration: 5000,
      continuous: false
    }
  },
  computed: {
    left () {
      if (!this.continuous ) {
        return false
      }
      return !this.reversed ? '0px' : 'auto'
    }
  },
  beforeDestroy () {
    this.clear()
  },
  methods: {
    clear () {
      clearInterval(this._timer)
      clearTimeout(this._throttle)
      this._timer = null
    },
    start () {
      this.clear()
      this.percent = 0
      this.reversed = false
      this.skipTimerCount = 0
      this.canSucceed = true

      if (this.throttle) {
        this._throttle = setTimeout(() => this.startTimer(), this.throttle)
      } else {
        this.startTimer()
      }
      return this
    },
    set (num) {
      this.show = true
      this.canSucceed = true
      this.percent = Math.min(100, Math.max(0, Math.floor(num)))
      return this
    },
    get () {
      return this.percent
    },
    increase (num) {
      this.percent = Math.min(100, Math.floor(this.percent + num))
      return this
    },
    decrease (num) {
      this.percent = Math.max(0, Math.floor(this.percent - num))
      return this
    },
    pause () {
      clearInterval(this._timer)
      return this
    },
    resume () {
      this.startTimer()
      return this
    },
    finish () {
      this.percent = this.reversed ? 0 : 100
      this.hide()
      return this
    },
    hide () {
      this.clear()
      setTimeout(() => {
        this.show = false
        this.$nextTick(() => {
          this.percent = 0
          this.reversed = false
        })
      }, 500)
      return this
    },
    fail (error) {
      this.canSucceed = false
      return this
    },
    startTimer () {
      if (!this.show) {
        this.show = true
      }
      if (typeof this._cut === 'undefined') {
        this._cut = 10000 / Math.floor(this.duration)
      }

      this._timer = setInterval(() => {
        /**
         * When reversing direction skip one timers
         * so 0, 100 are displayed for two iterations
         * also disable css width transitioning
         * which otherwise interferes and shows
         * a jojo effect
         */
        if (this.skipTimerCount > 0) {
          this.skipTimerCount--
          return
        }

        if (this.reversed) {
          this.decrease(this._cut)
        } else {
          this.increase(this._cut)
        }

        if (this.continuous) {
          if (this.percent >= 100) {
            this.skipTimerCount = 1

            this.reversed = !this.reversed
          } else if (this.percent <= 0) {
            this.skipTimerCount = 1

            this.reversed = !this.reversed
          }
        }
      }, 100)
    }
  },
}
</script>
<template>
  <div
    class="nuxt-progress"
    v-show="show"
    :class="{
      'nuxt-progress-notransition': skipTimerCount > 0,
      'nuxt-progress-failed': !canSucceed
    }"
    :style="{
      width: percent + '%',
      left: left
    }"
  ></div>
</template>
<style scoped>
.nuxt-progress {
  position: fixed;
  top: 0px;
  left: 0px;
  right: 0px;
  height: 2px;
  width: 0%;
  opacity: 1;
  transition: width 0.1s, opacity 0.4s;
  background-color: #F60102;
  z-index: 999999;
}

.nuxt-progress.nuxt-progress-notransition {
  transition: none;
}

.nuxt-progress-failed {
  background-color: #F60102;
}
</style>
```

**Reply by lacorde** [reply]
It works ! Thanks a lot.

**Reply by Gyurmatag** [reply]
Great, thanks! But it would be nice to have an official feature with documentation... Any news on this?

**Reply by andysay** [reply]
same

> Great, thanks! But it would be nice to have an official feature with documentation... Any news on this?

same question )

**Reply by andysay** [reply]
> I figure out it by fallowing code and nuxt-loading component ( refferanced from nuxt-2 ) in app.vue
> 
> **app.vue**
> 
> ```
> 
> <template>
>   <div class="scroll-smooth">
>     <NuxtPage/>
>     <nuxt-loading ref="loading" />
>   </div>
> </template>
> <script setup>
> 
>   const loading = ref(null)
> 
>   const nuxtApp = useNuxtApp()
> 
>   nuxtApp.hook("page:start", () => {
>     loading.value.start()
>   })
> 
>   nuxtApp.hook("page:finish", () => {
>     loading.value.finish()
>      window.scrollTo(0, 0)
>   })
> </script>
> ```
> 
> **nuxt-loading.vue**
> 
> ```
> 
> <script>
> export default {
>   name: 'NuxtLoading',
>   data () {
>     return {
>       percent: 0,
>       show: false,
>       canSucceed: true,
>       reversed: false,
>       skipTimerCount: 0,
>      
>       throttle: 200,
>       duration: 5000,
>       continuous: false
>     }
>   },
>   computed: {
>     left () {
>       if (!this.continuous ) {
>         return false
>       }
>       return !this.reversed ? '0px' : 'auto'
>     }
>   },
>   beforeDestroy () {
>     this.clear()
>   },
>   methods: {
>     clear () {
>       clearInterval(this._timer)
>       clearTimeout(this._throttle)
>       this._timer = null
>     },
>     start () {
>       this.clear()
>       this.percent = 0
>       this.reversed = false
>       this.skipTimerCount = 0
>       this.canSucceed = true
> 
>       if (this.throttle) {
>         this._throttle = setTimeout(() => this.startTimer(), this.throttle)
>       } else {
>         this.startTimer()
>       }
>       return this
>     },
>     set (num) {
>       this.show = true
>       this.canSucceed = true
>       this.percent = Math.min(100, Math.max(0, Math.floor(num)))
>       return this
>     },
>     get () {
>       return this.percent
>     },
>     increase (num) {
>       this.percent = Math.min(100, Math.floor(this.percent + num))
>       return this
>     },
>     decrease (num) {
>       this.percent = Math.max(0, Math.floor(this.percent - num))
>       return this
>     },
>     pause () {
>       clearInterval(this._timer)
>       return this
>     },
>     resume () {
>       this.startTimer()
>       return this
>     },
>     finish () {
>       this.percent = this.reversed ? 0 : 100
>       this.hide()
>       return this
>     },
>     hide () {
>       this.clear()
>       setTimeout(() => {
>         this.show = false
>         this.$nextTick(() => {
>           this.percent = 0
>           this.reversed = false
>         })
>       }, 500)
>       return this
>     },
>     fail (error) {
>       this.canSucceed = false
>       return this
>     },
>     startTimer () {
>       if (!this.show) {
>         this.show = true
>       }
>       if (typeof this._cut === 'undefined') {
>         this._cut = 10000 / Math.floor(this.duration)
>       }
> 
>       this._timer = setInterval(() => {
>         /**
>          * When reversing direction skip one timers
>          * so 0, 100 are displayed for two iterations
>          * also disable css width transitioning
>          * which otherwise interferes and shows
>          * a jojo effect
>          */
>         if (this.skipTimerCount > 0) {
>           this.skipTimerCount--
>           return
>         }
> 
>         if (this.reversed) {
>           this.decrease(this._cut)
>         } else {
>           this.increase(this._cut)
>         }
> 
>         if (this.continuous) {
>           if (this.percent >= 100) {
>             this.skipTimerCount = 1
> 
>             this.reversed = !this.reversed
>           } else if (this.percent <= 0) {
>             this.skipTimerCount = 1
> 
>             this.reversed = !this.reversed
>           }
>         }
>       }, 100)
>     }
>   },
> }
> </script>
> <template>
>   <div
>     class="nuxt-progress"
>     v-show="show"
>     :class="{
>       'nuxt-progress-notransition': skipTimerCount > 0,
>       'nuxt-progress-failed': !canSucceed
>     }"
>     :style="{
>       width: percent + '%',
>       left: left
>     }"
>   ></div>
> </template>
> <style scoped>
> .nuxt-progress {
>   position: fixed;
>   top: 0px;
>   left: 0px;
>   right: 0px;
>   height: 2px;
>   width: 0%;
>   opacity: 1;
>   transition: width 0.1s, opacity 0.4s;
>   background-color: #F60102;
>   z-index: 999999;
> }
> 
> .nuxt-progress.nuxt-progress-notransition {
>   transition: none;
> }
> 
> .nuxt-progress-failed {
>   background-color: #F60102;
> }
> </style>
> ```

Uncaught (in promise) TypeError: Cannot read properties of null (reading 'start') if app open at first time.

**Comment by wobsoriano** [comment]
Temporary module while waiting for official Nuxt implementation

https://github.com/wobsoriano/nuxt-progress

**Comment by shiva-prasad-reddy** [comment]
In SPA, i have auth-middleware which does a onetime auth on a page reload, so i have added a loading indicator using page:start and page:finish hooks, i found that the page hooks are being called after the middleware is run so facing a white-screen delay.
Is there anyway that i could add some html or loading component to body in middleware and remove it at the end of middleware.
